src/components/timeline/TrackScreenshots.js

```diff
@@ -300,6 +122,132 @@ class HoverPreview extends PureComponent<HoverPreviewProps> {
+  const [containerTop, setContainerTop] = useState<CssPixels | null>(null);
+
+  function handleMouseLeave() {
+    setPageX(null);
+    setOffsetX(null);
+    setContainerTop(null);
+  }
+
+  function handleMouseMove(event: SyntheticMouseEvent<HTMLDivElement>) {
+    const { top, left } = event.currentTarget.getBoundingClientRect();
+    setPageX(event.pageX);
+    setOffsetX(event.pageX - left);
+    setContainerTop(top);
+  }
+
+  // This selects a screenshot when clicking on the screenshot strip. Note that
+  // we use the mouseup event so that isMakingPreviewSelection is still
+  // accurate.
+  function selectScreenshotOnClick(event: SyntheticMouseEvent<HTMLDivElement>) {
+    if (isMakingPreviewSelection) {
+      // Avoid reseting the selection if the user is currently selecting one.
+      return;
+    }
+
+    const { left } = event.currentTarget.getBoundingClientRect();
+    const offsetX = event.pageX - left;
+    const screenshotIndex = findScreenshotAtMouse(offsetX);
+    if (screenshotIndex === null) return;
+    const { start, end } = screenshots[screenshotIndex];
+    if (end === null) return;
+    dispatch(
+      updatePreviewSelection({
+        hasSelection: true,
+        isModifying: false,
+        selectionStart: start,
+        selectionEnd: end,
+      })
+    );
+  }
+
+  function findScreenshotAtMouse(offsetX: number): number | null {
+    const rangeLength = rangeEnd - rangeStart;
+    const mouseTime = (offsetX / width) * rangeLength + rangeStart;
+
+    // Loop backwards to find the latest screenshot that has a time less
+    // than the current time at the mouse position.
+    for (let i = screenshots.length - 1; i >= 0; i--) {
+      const screenshotTime = screenshots[i].start;
+      if (mouseTime >= screenshotTime) {
+        return i;
+      }
+    }
+    return null;
+  }
+
+  let payload: ScreenshotPayload | null = null;
+
+  if (offsetX !== null) {
+    const screenshotIndex = findScreenshotAtMouse(offsetX);
+    if (screenshotIndex !== null) {
+      payload = (screenshots[screenshotIndex].data: any);
+    }
+  }
+
+  return (
+    <div
+      className="timelineTrackScreenshot"
+      ref={ref}
+      style={{ height: trackHeight }}
+      onMouseLeave={handleMouseLeave}
+      onMouseMove={handleMouseMove}
+      onMouseUp={selectScreenshotOnClick}
+    >
+      <ScreenshotStrip
+        thread={thread}
+        width={width}
+        rangeStart={rangeStart}
+        rangeEnd={rangeEnd}
+        screenshots={screenshots}
+        trackHeight={trackHeight}
+      />
+      {payload ? (
+        <HoverPreview
+          thread={thread}
+          isMakingPreviewSelection={isMakingPreviewSelection}
+          width={width}
+          pageX={pageX}
+          offsetX={offsetX}
+          containerTop={containerTop}
+          rangeEnd={rangeEnd}
+          rangeStart={rangeStart}
+          trackHeight={trackHeight}
+          payload={payload}
+        />
+      ) : null}
+    </div>
+  );
```

Comment:
Migrating to hooks is NOT fast, and care must be taken. But it almost completely worked on the first run once I got the types to pass. I ran into an issue where I was triggering infinite updates, and over-rendering for one thing. However, the component still worked the first load, which was a... weird feeling.

Reply:
I think hooks would be useful for some of the existing components. For example `useLayoutEffect` could be useful for these components where we use `requestAnimationFrame` before requesting things from the DOM (I'm thinking of Tooltips for example). There may be other good examples too.

Reply:
Another thing, is that any component that wants to use a hook MUST be done in the new style, this is a bit scary, as it means needing to add a trivial hook could trigger a full rewrite of a component.